networkx/algorithms/community/modularity_max.py

```diff
@@ -351,7 +353,20 @@ def greedy_modularity_communities(
 
     # continue merging communities until one of the breaking criteria is satisfied
     while len(communities) > cutoff:
-        dq = next(community_gen)
+        try:
+            dq = next(community_gen)
+        # StopIteration occurs when communities are the connected components
+        except StopIteration:
+            communities = sorted(communities, key=len, reverse=True)
+            # continue to merge even though modularity goes down?
```

Comment:
I have no idea what makes the most sense in this case, so I'd defer to the graph/network experts!

If no one's entirely sure what should happen here, one option could be to simply return the `communities` in the `best_n=None` (default) case and punt when `best_n!=None`, e.g. something like:

```python
# StopIteration occurs when communities are the connected components
except StopIteration:
    if best_n was None:  # Would need a variable to keep track of the original value of `best_n`
        return communities
    else:
        raise <exception-with-better-message-than-stop-iteration>
```
At least that would fix the returning-the-components regression part of the issue without having to necessarily make a decision on what to do with `best_n`.

Reply:
Luckily, the only value of `best_n` that would represent an `original_best_n is None` is when `best_n == len(G)`.
And in that case, the following while loop is passed over and we return `communities` just as in your suggestion.

I think the question mark in the comment on this line (361) should be removed -- and made more clear.
Once we include a parameter `best_n` we have to expect that sometimes we will merge communities that decrease the modularity. The parameter indicates to keep merging until we get `best_n` communities.
How about a change for that comment to:
```#  if best_n requires more merging, merge the biggest to keep modularity high```
and then can remove comment 3 lines down.